src/Agent.Worker/Build/BuildDirectoryManager.cs

```diff
@@ -144,6 +145,78 @@ public sealed class BuildDirectoryManager : AgentService, IBuildDirectoryManager
                     newConfig.RepositoryType = repository.Type;
                 }
 
+                // Make sure repository in the right location.
+                if (string.IsNullOrEmpty(newConfig.RepositoryDirectory))
+                {
+                    newConfig.RepositoryDirectory = newConfig.SourcesDirectory;
+                }
+
+                string requiredRepositoryPath;
+                var path = repository.Properties.Get<string>(RepositoryPropertyNames.Path);
+                if (!string.IsNullOrEmpty(path))
+                {
+                    if (path.IndexOfAny(Path.GetInvalidPathChars()) > -1)
+                    {
+                        throw new InvalidOperationException($"{path} contains invalid path characters.");
+                    }
+                    else if (Path.GetFileName(path).IndexOfAny(Path.GetInvalidFileNameChars()) > -1)
+                    {
+                        throw new InvalidOperationException($"{path} contains invalid folder name characters.");
+                    }
+                    else if (path.Contains($"..{Path.DirectorySeparatorChar}") || path.Contains($"..{Path.AltDirectorySeparatorChar}"))
+                    {
+                        throw new InvalidOperationException($"{path} can not contains relative path.");
+                    }
+                    else if (Path.IsPathRooted(path))
+                    {
+                        throw new InvalidOperationException($"{path} can not be a rooted path.");
+                    }
+                    else
+                    {
+                        requiredRepositoryPath = Path.Combine(newConfig.SourcesDirectory, path);
+                    }
+                }
+                else
+                {
+                    // When repository doesn't has path set, default to sources directory 1/s
+                    requiredRepositoryPath = newConfig.SourcesDirectory;
+                }
+
+                Trace.Info($"Repository requires to be placed at '{requiredRepositoryPath}', current location is '{newConfig.RepositoryDirectory}'");
+                if (!string.Equals(newConfig.RepositoryDirectory, requiredRepositoryPath, IOUtil.FilePathStringComparison))
+                {
+                    executionContext.Output($"Repository is current at '{newConfig.RepositoryDirectory}', move to '{requiredRepositoryPath}'.");
+                    var count = 1;
+                    var staging = Path.Combine(HostContext.GetDirectory(WellKnownDirectory.Work), newConfig.BuildDirectory, $"_{count}");
+                    while (Directory.Exists(staging))
```

Comment:
I'm not sure I understand this part of the code. Wouldn't this become an infinite loop if there's nothing else that deletes the staging folder?

Reply:
i am trying to find the folder name that doesn't exist yet.   staging_1/2/3/4/5/6....